File: GESDISC/AIRS.2023.03.24.240.L2.SUBS2SUP.v7.0.7.0.G23084133451.hdf.py
"""
This example code illustrates how to access and visualize a GESDISC AIRS swath
in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python AIRS.2023.03.24.240.L2.SUBS2SUP.v7.0.7.0.G23084133451.hdf.py

The HDF file must either be in your current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2023-04-24
"""

import os

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np

from mpl_toolkits.basemap import Basemap
from pyhdf.SD import SD, SDC

FILE_NAME = 'AIRS.2023.03.24.240.L2.SUBS2SUP.v7.0.7.0.G23084133451.hdf'
DATAFIELD_NAME = 'olr'

hdf = SD(FILE_NAME, SDC.READ)

# Read dataset.
data3D = hdf.select(DATAFIELD_NAME)
data = data3D[:,:]

# Read geolocation dataset.
lat = hdf.select('Latitude')
latitude = lat[:,:]
lon = hdf.select('Longitude')
longitude = lon[:,:]
        
    
# Replace the filled value with NaN, replace with a masked array.
data[data == -9999.0] = np.nan
datam = np.ma.masked_array(data, np.isnan(data))
    
m = Basemap(projection='cyl', resolution='l',
            llcrnrlat=-90, urcrnrlat = 90,
            llcrnrlon=-180, urcrnrlon = 180)
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(-90, 91, 30), labels=[True,False,False,False])
m.drawmeridians(np.arange(-180, 180, 45), labels=[False,False,False,True])
sc = m.scatter(longitude, latitude, c=data, s=0.1, cmap=plt.cm.jet,
                edgecolors=None, linewidth=0)
cb = m.colorbar()
# See [1]. The dataset doesn't have unit attribute.
units = 'W/m^2'
cb.set_label(units)
    
basename = os.path.basename(FILE_NAME)
plt.title('{0}\n {1}'.format(basename, DATAFIELD_NAME))
fig = plt.gcf()
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)
    
# Reference
#
# [1] https://airs.jpl.nasa.gov/data/products/v7-L2-L3/

File: GESDISC/AIRS_2023_03_24_240_L2_SUBS2SUP_v7_0_7_0_G23084133451_hdf.m
%
% This example code illustrates how to access, merge, subset, and
% visualize GES DISC AIRS L2 swath files in MATLAB. 
%
%  If you have any questions, suggestions, comments  on this
%  example, please use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%
%   If you would like to see an  example of any other NASA HDF/HDF-EOS data
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org 
% or post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
% $matlab -nosplash -nodesktop -r AIRS_2023_03_24_240_L2_SUBS2SUP_v7_0_7_0_G23084133451_hdf
%
% Tested under: MATLAB R2023a
% Last updated: 2023-05-03

import matlab.io.hdfeos.*
import matlab.io.hdf4.*


FILE_NAME = 'AIRS.2023.03.24.240.L2.SUBS2SUP.v7.0.7.0.G23084133451.hdf';
DATAFIELD_NAME = 'olr';

% Read attributes from the data field.
SD_id = sd.start(FILE_NAME, 'rdonly');
sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
sds_id = sd.select(SD_id, sds_index);


% Read _FillValue from the data field.
fillvalue_index = sd.findAttr(sds_id, '_FillValue');
fillvalue = sd.readAttr(sds_id, fillvalue_index);

% Terminate access to the corresponding data set.
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);


f = figure('Name', FILE_NAME, 'visible', 'off');


SWATH_NAME='L2_Support_atmospheric&surface_product';
    
% Open HDF-EOS2 file.
file_id = sw.open(FILE_NAME, 'rdonly');

% Open swath.
swath_id = sw.attach(file_id, SWATH_NAME);
data = sw.readField(swath_id, DATAFIELD_NAME, [], [], []);
    
% Read lat and lon data.
lon = sw.readField(swath_id, 'Longitude', [], [], []);
lat = sw.readField(swath_id, 'Latitude', [], [], []);

% Print whether it's asceding or descending.
attr = sw.readAttr(swath_id, 'node_type')

% Convert the data to double type for plot.
data=double(data);
lon=double(lon);
lat=double(lat);

% Replace the filled value with NaN.
data(data==fillvalue) = NaN;

% Detach from the Swath object.
sw.detach(swath_id);
sw.close(file_id);


% Create the plot.
axesm('MapProjection','eqdcylin',...
      'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel','south')

coast = load('coastlines.mat');

scatterm(lat(:), lon(:), 1, data(:));

colormap('Jet');
h=colorbar();
plotm(coast.coastlat, coast.coastlon, 'k');


% Draw unit.
set(get(h, 'title'), 'string', 'None', ...
                  'FontSize', 12, 'FontWeight','bold', ...
                  'Interpreter', 'none');

% Put title.
tstring = {FILE_NAME;DATAFIELD_NAME};
title(tstring, 'Interpreter', 'none', 'FontSize', 16, ...
      'FontWeight','bold');


% The following fixed-size screen size will look better in JPEG if
% your screen is too large.
scrsz = [1 1 800 600];
set(f, 'position', scrsz, 'PaperPositionMode', 'auto');
tightmap;
saveas(f, [FILE_NAME '.m.png']);
exit;
